File: apps/desktop/src/ai/localRuntime.ts
import { invoke, isTauri } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";
import { readLocalAiSettings, worldWanBudget } from "../domain/localAiSettings";

export type LocalRuntimeFeature = "characterPixal3d" | "characterTrellis2" | "characterRig" | "propImageLite" | "propImageKrea" | "world";

export interface LocalRuntimeStatus {
  state: "ready" | "external" | "needsStart" | "needsInstall";
  endpoint: string;
  feature: LocalRuntimeFeature;
  installedBytes: number;
  totalBytes: number;
  requiredBytes: number;
  message: string;
}

export interface LocalRuntimeProgress {
  feature: LocalRuntimeFeature;
  stage: "runtime" | "extracting" | "downloading" | "verifying" | "nodes" | "dependencies" | "trainer" | "starting" | "ready" | "training";
  completedBytes: number;
  totalBytes: number;
  message: string;
}

export interface TrainedWorld {
  path: string;
  filename: string;
}

export const formatRuntimeBytes = (bytes: number): string => bytes >= 1024 ** 3
  ? `${(bytes / 1024 ** 3).toFixed(1)} GB`
  : `${Math.max(1, Math.ceil(bytes / 1024 ** 2))} MB`;

export const runtimeProgressPercent = (progress: Pick<LocalRuntimeProgress, "completedBytes" | "totalBytes"> & Partial<Pick<LocalRuntimeProgress, "stage">>): number => {
  if (progress.stage === "ready") return 100;
  if (progress.totalBytes <= 0) return 0;
  // Download bytes can be complete while extraction, dependency installation,
  // verification, or server startup is still running. Reserve 100% for the
  // native ready event so the bar never promises completion prematurely.
  return Math.min(99, Math.max(0, Math.round(progress.completedBytes / progress.totalBytes * 100)));
};

export async function getLocalRuntimeStatus(feature: LocalRuntimeFeature): Promise<LocalRuntimeStatus | null> {
  if (!isTauri()) return null;
  return await invoke<LocalRuntimeStatus>("local_runtime_status", { feature });
}

export async function ensureLocalRuntime(feature: LocalRuntimeFeature, onProgress?: (progress: LocalRuntimeProgress) => void): Promise<LocalRuntimeStatus> {
  if (!isTauri()) throw new Error("Automatic local creation setup is available in the installed DnDRom desktop app.");
  const unlisten = await listen<LocalRuntimeProgress>("local-runtime-progress", ({ payload }) => {
    if (payload.feature === feature) onProgress?.(payload);
  });
  try {
    const budget = worldWanBudget(readLocalAiSettings().memoryProfile);
    return await invoke<LocalRuntimeStatus>("ensure_local_runtime", { feature, vramReserveGb: budget.vramReserveGb });
  } finally {
    unlisten();
  }
}

export async function restartLocalRuntime(feature: LocalRuntimeFeature, onProgress?: (progress: LocalRuntimeProgress) => void): Promise<LocalRuntimeStatus> {
  if (!isTauri()) throw new Error("Automatic local creation recovery is available in the installed DnDRom desktop app.");
  const unlisten = await listen<LocalRuntimeProgress>("local-runtime-progress", ({ payload }) => {
    if (payload.feature === feature) onProgress?.(payload);
  });
  try {
    const budget = worldWanBudget(readLocalAiSettings().memoryProfile);
    return await invoke<LocalRuntimeStatus>("restart_local_runtime", { feature, vramReserveGb: budget.vramReserveGb });
  } finally {
    unlisten();
  }
}

export async function provisionLocalCreationSuite(onProgress?: (progress: LocalRuntimeProgress) => void): Promise<LocalRuntimeStatus | null> {
  if (!isTauri()) return null;
  const unlisten = await listen<LocalRuntimeProgress>("local-runtime-progress", ({ payload }) => onProgress?.(payload));
  try {
    const budget = worldWanBudget(readLocalAiSettings().memoryProfile);
    return await invoke<LocalRuntimeStatus>("provision_local_creation_suite", { vramReserveGb: budget.vramReserveGb });
  } finally {
    unlisten();
  }
}

export async function trainLocalWorld(datasetPath: string, onProgress?: (progress: LocalRuntimeProgress) => void): Promise<TrainedWorld> {
  const unlisten = await listen<LocalRuntimeProgress>("local-runtime-progress", ({ payload }) => {
    if (payload.feature === "world") onProgress?.(payload);
  });
  try {
    const budget = worldWanBudget(readLocalAiSettings().memoryProfile);
    return await invoke<TrainedWorld>("train_local_world", {
      datasetPath,
      totalSteps: budget.trainingSteps,
      maxSplats: budget.maxSplats,
      maxFrames: budget.trainingMaxFrames,
    });
  } finally {
    unlisten();
  }
}

/** Returns only a recently completed, structurally validated COLMAP dataset
 * inside DnDRom's managed runtime. Used after the UI misses Comfy's path text. */
export async function findLatestLocalWorldDataset(): Promise<string | null> {
  if (!isTauri()) return null;
  return await invoke<string | null>("find_latest_local_world_dataset");
}

/** Recovers Brush output that completed before a late validation/import error,
 * allowing Retry to resume at 98% rather than training the same dataset again. */
export async function findLatestLocalTrainedWorld(): Promise<TrainedWorld | null> {
  if (!isTauri()) return null;
  return await invoke<TrainedWorld | null>("find_latest_local_trained_world");
}

export async function readLocalRuntimeFile(path: string, filename: string): Promise<File> {
  const response = await invoke<ArrayBuffer | Uint8Array | number[]>("read_local_runtime_file", { path });
  const bytes = response instanceof ArrayBuffer ? new Uint8Array(response) : response instanceof Uint8Array ? response : new Uint8Array(response);
  const owned = new Uint8Array(bytes.byteLength);
  owned.set(bytes);
  return new File([owned.buffer], filename, { type: "application/octet-stream" });
}

File: apps/desktop/src-tauri/src/local_runtime.rs
//! Tauri transport adapter; model ownership lives in dndrom-local-runtime.
pub use dndrom_local_runtime::{LocalRuntimeState, RuntimeFeature, RuntimeStatus, TrainedWorld};
use tauri::{AppHandle, Emitter, Manager, State};

pub fn context(app: &AppHandle) -> Result<dndrom_local_runtime::RuntimeContext, String> {
    let data_dir = app.path().app_local_data_dir().map_err(|e| e.to_string())?;
    let events = app.clone();
    Ok(dndrom_local_runtime::RuntimeContext::new(
        data_dir,
        move |progress| {
            let _ = events.emit("local-runtime-progress", progress);
        },
    ))
}
#[tauri::command]
pub async fn local_runtime_status(
    app: AppHandle,
    feature: RuntimeFeature,
) -> Result<RuntimeStatus, String> {
    let ctx = context(&app)?;
    tauri::async_runtime::spawn_blocking(move || {
        dndrom_local_runtime::local_runtime_status(ctx, feature)
    })
    .await
    .map_err(|e| e.to_string())?
}
#[tauri::command]
pub async fn ensure_local_runtime(
    app: AppHandle,
    state: State<'_, LocalRuntimeState>,
    feature: RuntimeFeature,
    vram_reserve_gb: Option<f32>,
) -> Result<RuntimeStatus, String> {
    let ctx = context(&app)?;
    let manager = state.inner().clone();
    tauri::async_runtime::spawn_blocking(move || {
        dndrom_local_runtime::ensure_local_runtime(ctx, &manager, feature, vram_reserve_gb)
    })
    .await
    .map_err(|e| e.to_string())?
}
#[tauri::command]
pub async fn restart_local_runtime(
    app: AppHandle,
    state: State<'_, LocalRuntimeState>,
    feature: RuntimeFeature,
    vram_reserve_gb: Option<f32>,
) -> Result<RuntimeStatus, String> {
    let ctx = context(&app)?;
    let manager = state.inner().clone();
    tauri::async_runtime::spawn_blocking(move || {
        dndrom_local_runtime::restart_local_runtime(ctx, &manager, feature, vram_reserve_gb)
    })
    .await
    .map_err(|e| e.to_string())?
}
#[tauri::command]
pub async fn provision_local_creation_suite(
    app: AppHandle,
    state: State<'_, LocalRuntimeState>,
    vram_reserve_gb: Option<f32>,
) -> Result<RuntimeStatus, String> {
    let ctx = context(&app)?;
    let manager = state.inner().clone();
    tauri::async_runtime::spawn_blocking(move || {
        dndrom_local_runtime::provision_local_creation_suite(ctx, &manager, vram_reserve_gb)
    })
    .await
    .map_err(|e| e.to_string())?
}
#[tauri::command]
pub fn find_latest_local_world_dataset(app: AppHandle) -> Result<Option<String>, String> {
    dndrom_local_runtime::find_latest_local_world_dataset(context(&app)?)
}
#[tauri::command]
pub fn find_latest_local_trained_world(app: AppHandle) -> Result<Option<TrainedWorld>, String> {
    dndrom_local_runtime::find_latest_local_trained_world(context(&app)?)
}
#[tauri::command]
pub async fn train_local_world(
    app: AppHandle,
    state: State<'_, LocalRuntimeState>,
    dataset_path: String,
    total_steps: Option<u32>,
    max_splats: Option<u32>,
    max_frames: Option<u32>,
) -> Result<TrainedWorld, String> {
    let ctx = context(&app)?;
    let manager = state.inner().clone();
    tauri::async_runtime::spawn_blocking(move || {
        dndrom_local_runtime::train_local_world(
            ctx,
            &manager,
            dataset_path,
            total_steps,
            max_splats,
            max_frames,
        )
    })
    .await
    .map_err(|e| e.to_string())?
}
#[tauri::command]
pub fn read_local_runtime_file(
    app: AppHandle,
    path: String,
) -> Result<tauri::ipc::Response, String> {
    dndrom_local_runtime::read_local_runtime_file(context(&app)?, path)
        .map(tauri::ipc::Response::new)
}
